### PR TITLE
Geocode Literal proposal

### LINKED ISSUE #444: Geocoding Support in GeoSPARQL
Our group has already received a request from @paulc-dstl to add GeoHash literals to GeoSPARQL.

However, there are more geocoding systems which could be taken into account here.

In fact, a lot of geocoding systems can be resolved to coordinates in either UTM or WGS84-based coordinate reference systems.

My question here:

Would it make sense to maybe define two literal types: "wgs84GeocodingLiteral" and "utmGeoCodingLiteral" and indicate the Geocoding system with a URI definition, similar to how we do it in WKT?
We could curate a list of URIs for well-known geocoding systems which implementers could use.

Or should we proceed with defining GeoHash, and people should create their own literal types outside the standard?

### PR DESCRIPTION
**Pull request by situx**
Proposal of Geocode Literal Type for me and @paulc-dstl to jointly work on

Closes #444 #428 

- [x] Spec updated
- [x] Ontology updated
- [x] Functions updated
- [x] Update SHACL shapes?

### CODE REVIEW

File: spec/sections/11-geometry_extension.adoc [hunk 1]
```diff
@@ -793,6 +793,105 @@ The function http://www.opengis.net/def/function/geosparql/asKML[`geof:asKML`] c
...
 
+
+==== Geocode Literal
+
+This section establishes the Requirements class for representing geocode data in RDF based on a Literal type that can encompass different Geocodes. It defines one RDFS Datatype:
+<<RDFS Datatype: geo:geocodeLiteral, Geocode Literal>> and one property, <<Function: geof:asGeocode, as GeoCode>>.
+
+===== RDFS Datatype: geo:geocodeLiteral
+
+The datatype http://www.opengis.net/ont/geosparql#kmlLiteral[`geo:geocodeLiteral`] is used to contain a geocode serialization of a Geometry.
```
Comment (mperry455) [inline spec/sections/11-geometry_extension.adoc:804]: Is kmlLiteral the correct link here?

File: spec/sections/11-geometry_extension.adoc [hunk 2]
```diff
@@ -793,6 +793,105 @@ The function http://www.opengis.net/def/function/geosparql/asKML[`geof:asKML`] c
...
+====
+RDFS Literals of type <<RDFS Datatype: geo:geocodeLiteral, `geo:geocodeLiteral`>> contain a Geocoding system definition in the form of an IRI. The SRS for the All literals of this type shall according to the KML specification only be encoded in and assumed to use the WGS84 geodetic longitude-latitude spatial reference system (http://www.opengis.net/def/crs/OGC/1.3/CRS84[`http://www.opengis.net/def/crs/OGC/1.3/CRS84`]).
+====
+
+The example <<RDFS Datatype: geo:geocodeLiteral, Geocode Literal>> below encodes a point Geometry (point 27 19' 32S, 153 4' 59"E) in the GeoHash encoding:
+
+[%unnumbered]
+[source,turtle]
+----
+"<https://w3id.org/geohashes/geohash> r7hue3x9jepv"^^geo:geoHasLiteral"^^<http://www.opengis.net/ont/geosparql#geocodeLiteral>
```
Comment (mperry455) [inline spec/sections/11-geometry_extension.adoc:840]: ^^geo:geoHasLiteral" should be removed

File: spec/sections/11-geometry_extension.adoc [hunk 3]
```diff
@@ -793,6 +793,105 @@ The function http://www.opengis.net/def/function/geosparql/asKML[`geof:asKML`] c
...
+====
+Implementations shall allow the RDF property <<Property: geo:asGeocode, `geo:asGeocode`>> to be used in SPARQL graph patterns.
+====
+
+The property <<Property: geo:asGeocode, as Geocode>> is used to link a geometric element with a Geocode serialization.
+
+[%unnumbered]
+[source,turtle]
+----
+geo:asKML
```
Comment (mperry455) [inline spec/sections/11-geometry_extension.adoc:866]: Should be geo:asGeocode

File: spec/ontology/vocabularies/funcsrules.ttl [hunk 4]
```diff
@@ -662,14 +662,42 @@ funcs:asKML
...
+
+funcs:asGeocode_param1 a fno:Parameter ;
+    fno:type geo:wktLiteral, geo:gmlLiteral, geo:geoJSONLiteral, geo:kmlLiteral, geo:dggsLiteral, geo:geocodeLiteral ;
+    fno:required "true"^^xsd:boolean ;
+    skos:prefLabel "geom"@en .
+
+funcs:asGeocode_param2 a fno:Parameter ;
+    fno:type rdfs:Resource ;
+    fno:required "true"^^xsd:boolean ;
+    skos:prefLabel "geom"@en .
```
Comment (mperry455) [inline spec/ontology/vocabularies/funcsrules.ttl:694]: prefLabel should be "geocode"@en

**Review by mperry455 — APPROVED** [review]
Looks good except for a few copy/paste errors

**Comment by situx** [comment]
Added all changes requested by @mperry455

**Comment by paulc-dstl** [comment]
Looks good to me. A neat solution to support all types of geocoded serialisations. 

Regarding the original Change Request, I'm not sure it will satisfy the needs of the IES devs as the intent there is to use geocoded literals in IRI patterns; this GeoSPARQL pattern neither helps nor hinders that. But I think that is fine so would consider the Change Request complete, any IRI matters will need to be dealt with separately. 

Thanks!

**Comment by situx** [comment]
@paulc-dstl Could you give an example illustrating what the IES devs would like to see implemented?

**Review by nicholascar — APPROVED** [review]

**Comment by nicholascar** [comment]
One think missing here is how/where to get the IRIs that identify geocode types from?

This is akin to getting IRIs for DGGSes and even CRSes. CRSes we have a couple of major registries - epsg.io & OGC - but there is nothing yet for CRS or geocode.

Could the OGC manage a (very short) list of geogode type identifiers? This is probably a question for the Geosemantics DWG @jabhay @lvdbrink

**Comment by situx** [comment]
I added the missing SHACL shapes. Now the PR is ready unless we want to provide a comment on where to find URIs for Geocodes.

**Pull request merged**